File: backend/src/models/BillItem.js
const mongoose = require('mongoose');

const billItemSchema = new mongoose.Schema({
  billId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Bill',
    required: true,
    index: true,
  },
  productId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Product',
    default: null,
  },
  productName: {
    type: String,
    required: true,
  },
  productNameHindi: {
    type: String,
    default: '',
  },
  unit: {
    type: String,
    required: true,
  },
  quantity: {
    type: Number,
    required: true,
    min: 0,
  },
  defaultRate: {
    type: Number,
    default: 0,
  },
  appliedRate: {
    type: Number,
    required: true,
  },
  amount: {
    type: Number,
    required: true,
  },
}, { timestamps: { createdAt: true, updatedAt: false } });


billItemSchema.index({ productId: 1 });

module.exports = mongoose.model('BillItem', billItemSchema);

File: lib/models/bill_item.dart
import 'package:equatable/equatable.dart';

class BillItem extends Equatable {
  final String id;
  final String billId;
  final String? productId;
  final String productName;
  final String unit;
  final double quantity;
  final double defaultRate;
  final double appliedRate;
  final double amount;

  const BillItem({
    required this.id,
    required this.billId,
    this.productId,
    required this.productName,
    required this.unit,
    required this.quantity,
    this.defaultRate = 0,
    required this.appliedRate,
    required this.amount,
  });

  factory BillItem.fromJson(Map<String, dynamic> json) => BillItem(
    id: json['_id'] ?? json['id'] ?? '',
    billId: json['billId'] ?? '',
    productId: json['productId'],
    productName: json['productName'] ?? '',
    unit: json['unit'] ?? 'kg',
    quantity: (json['quantity'] ?? 0).toDouble(),
    defaultRate: (json['defaultRate'] ?? 0).toDouble(),
    appliedRate: (json['appliedRate'] ?? 0).toDouble(),
    amount: (json['amount'] ?? 0).toDouble(),
  );

  Map<String, dynamic> toJson() => {
    if (productId != null) 'productId': productId,
    'productName': productName,
    'unit': unit,
    'quantity': quantity,
    'defaultRate': defaultRate,
    'appliedRate': appliedRate,
  };

  @override
  List<Object?> get props => [id, productName, quantity, appliedRate, amount];
}
